docs/design/2018-09-04-histograms-in-plan.md

```diff
@@ -0,0 +1,144 @@
+- Since the boundary of the bucket have been aligned. Then we can estimate the join's size bucket by bucket by the well-know way.
+
+Original:
+
+<img alt="original histogram" src="./imgs/histogram-1.jpeg" width="500pt"/>
+
+Step 1:
+
+<img alt="Step 1" src="./imgs/histogram-2.jpeg" width="500pt"/>
+
+Step 2:
+
+<div style="text-align: center">
+<img alt="Step 2" src="./imgs/histogram-3.png" width="150pt"/>
+</div>
+
+The calculation inside the bucket can be calculated as this formula <img alt="$selectivity=joinKeySelectivity*RowCount(t1)*RowCount(t2)$" src="imgs/histogram-formula-1.png?invert_in_darkmode" align="middle" width="476.19356895pt" height="24.657534pt"/>
+
+Where <img alt="$joinKeySelectivity = \frac{1}{NDV(t1.key)}*\frac{1}{NDV(t2.key)}*ndvAfterJoin$" src="imgs/histogram-formula-2.png?invert_in_darkmode" align="middle" width="466.6166208pt" height="27.7756545pt"/>.
+
+The `ndvAfterJoin` can be <img alt="$min(NDV(t1.key), NDV(t2.key))$" src="imgs/histogram-formula-3.png?invert_in_darkmode" align="middle" width="248.4423216pt" height="24.657534pt"/> or a more detailed one if we can calculate it.
+
+
+Since it won’t have one side filter, we only need to consider the composite filters after considering the join key. We can simply multiply `selectionFactor` if there are other composite filters since `Selectivity` cannot calculate selectivity of an expression containing multiple columns.
+
+##### One side outer join
+It's almost the same as inner join's behavior. But we need to consider two more thing:
+
+- The unmatched row will be filled as `NULL`. This should be calculated in the new histogram. The null count can be calculated when we estimate the matched count bucket by bucket.
+- There will be one side filters of the outer table. If the filter is about join key and can be converted to range information, it's can be easily involved when we do the calculation bucket by bucket. Otherwise it's a little hard to deal with it. Don't consider this case currently.
+
+##### Semi join
+It’s something similar to inner join. But no data expanding occurs. So the `total` can be adjusted easily.
+
+##### Anti semi join
+Same with semi join.
+
+#### `Aggregate`
+Just read the NDV information from the `statsInfo` to decide the row count after aggregate. If there's index can fully match the group-by items, we will just use its NDV. Otherwise we will multiply the ndv of each column(or index that can match part of the group-by item).
+
+If some of the group-by items are also in the select field, we will create new histograms by modifying the `totalCnt` of each bucket(set it the same with `NDV`).
```

Comment:
Could you please give an example of this?

Reply:
Before aggregation, we have a histogram whose buckets are `range: [1, 10], ndv:3, cnt: 10`, `range [20, 25], ndv: 5, cnt: 100`.
If this column is group-by item and appears in select field, we will have a new histogram whose buckets are `range: [1, 10], ndv:3, cnt:3`, `range [20, 25], ndv: 5, cnt: 5`.